### PR TITLE
[GraphExecutor] Shared Memory bind and runtime reshaping

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

**Comment by mli** [comment]
Can one of the admins verify this patch?

File: include/mxnet/symbolic.h [hunk 1]
```diff
@@ -351,7 +352,8 @@ class Executor {
                         const std::vector<NDArray> &in_args,
                         const std::vector<NDArray> &arg_grad_store,
                         const std::vector<OpReqType> &grad_req_type,
-                        const std::vector<NDArray> &aux_states);
+                        const std::vector<NDArray> &aux_states,
+                        Executor* shared_exec = NULL);
```
Comment (tqchen) [inline include/mxnet/symbolic.h]: nullptr

**Comment by tqchen** [comment]
Please add a few test cases on this, also suggest the following improvments.

- Make executor take a shared_ptr of the StorageAllocator (this helps cleanup the code reference issue).

- Being able to call release and re-aquire space during runtime, so an executor can optionally release its requested memory, and re-aquire it during imperative execution (optional, can do it later).

**Comment by piiswrong** [comment]
I thought about sharing StorageAllocator at the start, but that actually doesn't work. Suppose ExecA has 2 chunks of memory, Rebind ExecB reuse 1 chunk and allocate 1 new chunk. Then ExecA destructs. Now The un-reused 1 chunk of memory stays there until all derived executors are deleted.

**Comment by tqchen** [comment]
Understand this case, however, I guess the point of shared memory execs is that all the shared execs destructs together. We can discuss a bit on this offline

**Comment by piiswrong** [comment]
Not really. Usually you have one base exec. Then for each batch you derive a new exec with different size. After this batch the new exec destructs. Then if it allocated any new memory they will stay there unused before base destructs.

**Comment by piiswrong** [comment]
@tqchen Done. Lets do a final pass and merge this.

File: src/symbol/graph_memory_allocator.h [hunk 2]
```diff
@@ -164,7 +196,9 @@ GraphStorageAllocator::Request(Context ctx, int type_flag, TShape shape, uint32_
     StorageEntry *e = it->second;
     if (e->ctx != ctx) continue;
     if (e->type_flag != type_flag) continue;
-    if (node_color_[e->released_by_node] != node_color_[node_id]) continue;
+    if (node_color_[e->released_by_node] != num_match_color_ + 1
```
Comment (tqchen) [inline src/symbol/graph_memory_allocator.h]: make dummy color a constant, so it is easier to expect what is going on

**Pull request merged**